src/Microsoft.Extensions.Http/Logging/LoggingHttpMessageHandler.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Http.Logging
+{
+    public class LoggingHttpMessageHandler : DelegatingHandler
+    {
+        private ILogger _logger;
+
+        public LoggingHttpMessageHandler(ILogger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _logger = logger;
+        }
+
+        protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            // Not using a scope here because we always expect this to be at the end of the pipeline, thus there's
+            // not really anything to surround.
+            Log.RequestStart(_logger, request);
```

Comment:
I really prefer this to the extension methods that are common. This way you don't accidentally use logging that's intended for another component. We've legitimately had bugs like this 4-5 times in MVC.

Reply:
Does anyone mind?

Reply:
> This way you don't accidentally use logging that's intended for another component.

Interesting. Is that because the extension methods were public? Would making them internal just solve that?

Reply:
Another thought I had when doing this somewhere else was putting the extensions on `ILogger<T>` (for a specific `T`). Doesn't necessarily help you here though.

This seems OK to me. It'd be better if it could still be an extension method but C# doesn't allow that.

Reply:
> Would making them internal just solve that?

That would only keep you from crossing assembly boundaries. There are often multiple categories of `ILogger` within an assembly, which gets kinda gross.

Reply:
> Is that because the extension methods were public? Would making them internal just solve that?

No, the problem is about other components inside the assembly.